File: src/Scripting/PHP/decorator.php
<?php

declare(strict_types=1);

interface Component
{
    public function render(): string;
}

final class PlainMessage implements Component
{
    public function render(): string { return 'alert'; }
}

abstract class ComponentDecorator implements Component
{
    public function __construct(protected Component $inner) {}
}

final class AuditDecorator extends ComponentDecorator
{
    public function render(): string { return 'audit(' . $this->inner->render() . ')'; }
}

final class EncryptDecorator extends ComponentDecorator
{
    public function render(): string { return 'enc(' . $this->inner->render() . ')'; }
}

$base = new PlainMessage();
echo 'base=' . $base->render() . PHP_EOL;
echo 'audit=' . (new AuditDecorator($base))->render() . PHP_EOL;
echo 'encrypted=' . (new EncryptDecorator($base))->render() . PHP_EOL;
echo 'stacked=' . (new AuditDecorator(new EncryptDecorator($base)))->render() . PHP_EOL;

File: src/DataScience/Octave/decorator.m
function decorator()
  base = @base_render;
  audited = audit_decorator(base);
  encrypted = encrypt_decorator(base);
  stacked = audit_decorator(encrypt_decorator(base));

  fprintf('base=%s\n', base());
  fprintf('audit=%s\n', audited());
  fprintf('encrypted=%s\n', encrypted());
  fprintf('stacked=%s\n', stacked());
end

function output = base_render()
  output = 'alert';
end

function wrapped = audit_decorator(component)
  wrapped = @render;
  function output = render()
    output = ['audit(' component() ')'];
  end
end

function wrapped = encrypt_decorator(component)
  wrapped = @render;
  function output = render()
    output = ['enc(' component() ')'];
  end
end
